packages/arb-token-bridge-ui/src/hooks/useAllowance.ts

```diff
@@ -0,0 +1,142 @@
+import { useCallback, useEffect, useMemo, useState } from 'react'
+import { useAccount } from 'wagmi'
+import useSWR from 'swr'
+import { ERC20__factory } from '@arbitrum/sdk/dist/lib/abi/factories/ERC20__factory'
+
+import { useNetworksAndSigners } from './useNetworksAndSigners'
+import {
+  getL2ERC20Address,
+  getL2GatewayAddress,
+  getL1TokenAllowance
+} from '../util/TokenUtils'
+
+export const useAllowance = (erc20L1Address: string | undefined) => {
+  const {
+    l1: { provider: l1Provider },
+    l2: { provider: l2Provider }
+  } = useNetworksAndSigners()
+  const { address } = useAccount()
+
+  const [erc20L2Address, setErc20L2Address] = useState<string | undefined>()
+
+  useEffect(() => {
+    async function updateErc20L2Address() {
+      setErc20L2Address(
+        erc20L1Address
+          ? await getL2ERC20Address({
+              erc20L1Address,
+              l1Provider,
+              l2Provider
+            })
+          : undefined
+      )
+    }
+    updateErc20L2Address()
+  }, [erc20L1Address, l1Provider, l2Provider])
+
+  // Query keys
+
+  const isValidQueryKey = !!(
+    address &&
+    erc20L1Address &&
+    l1Provider &&
+    l2Provider
+  )
+
+  const queryKeyL1 = useMemo(() => {
+    if (!isValidQueryKey) {
+      // don't fetch
+      return undefined
+    }
+    return 'allowanceL1'
```

Comment:
the `key` for SWR should include all parameters passed to the fetcher (https://swr.vercel.app/docs/arguments)